### ISSUE TITLE
incognito vs normal mode

### ISSUE THREAD

**Issue by nekromoff** [EFForg/privacybadgerfirefox#437]
badger does not respect incognito mode and keeps the same settings enabled/disabled in incognito for the normal mode as well. possible privacy breach.

**Comment by cooperq** [comment]
@nekromoff if I understand you correctly you are talking about settings that the user manually overrides, yes? Such as manually moving a domain to Red, Yellow, or Green? I think there is a bit of a debate about what the expected behavior might be here.
On one hand the user might expect that any settings made by the user will stay the same whether or not they are in incognito mode, whereas another user might have your expectations that changes made in incognito mode will not persist. 
Perhaps an optimal solution to this would be to warn users browsing in incognito mode that changes made to the settings in incognito mode will persist.

**Comment by nekromoff** [comment]
Yes, exactly, that was my point. I expect Incognito mode to be as pure
browser as it gets, e.g. all default settings used, extensions included.

But you are correct that some users might want to keep the behaviour the
same. Maybe an option in preferences for this?

**Comment by cooperq** [comment]
Are you saying that you would also expect privacy badger to bring none of your settings from normal browsing mode into incognito mode when you start a new incognito session? i.e. privacy badger would start with a fresh database and have to figure out what to block all over again when you entered incognito mode?

**Comment by nekromoff** [comment]
Yes, that's what I am saying. Since no settings should be kept in Incognito
mode, I don't see why Badger should make a difference.

**Comment by nekromoff** [comment]
Sorry, I meant no settings should be kept *after closing Incognito* (e.g.
between Incognito old and new sessions).

**Comment by cooperq** [comment]
okay so you are not saying that normal browser settings should not transfer over to incognito mode, you are saying that changes in incognito mode should not persist back to normal mode (which I agree with), is that correct?

**Comment by nekromoff** [comment]
Well, I mean it both ways - Incognito mode is kind of sandbox mode, is it
not?

**Comment by cooperq** [comment]
I respectfully disagree, incognito mode essentially means 'never remember history and delete all cookies and identifiers when finished', you still get your bookmarks, saved passwords, settings, etc. 
I think that the proper behavior here is for privacy badger to continue doing as it does but for us to notify the user that setting changes made in incognito mode will persist into normal mode.

**Comment by antistress** [comment]
I agree with Cooper Quintin
In incognito mode I can see changes made into my main profile (bookmarks, history and so on) whereas changes made in incognito mode don't appear into my main profile.
But you have to decide which is default behaviour in incognito mode : don't save settings made in this mode or do save settings in this mode. In both case, user should be warn

**Comment by antistress** [comment]
To decide which is default behaviour in incognito mode (whether don't save settings made in this mode or do save settings in this mode), you should wander if Privacy Badger settings may reveal sites that have been visited. If yes, then default behaviour should be to not save settings in Incognito mode (and to warn the user of that), or PB may "corrupt" incognito mode. If no, then default behaviour could be to save settings changes made in Incognito mode (and to warn the user of that).

**Comment by nekromoff** [comment]
I agree with @antistress.

**Comment by jawz101** [comment]
Incognito mode should offer more protection than browsing without Privacy Badger enabled.  I think it should carry the rules over from browsing non-Incognito as a read-only list but not start from scratch.

Incognito was initially to protect yourself from other people who could check out your browser history client-side but it should also help you to browse without being tracked by server-side 3rd parties.

**Comment by tshanks** [comment]
I concur with ghost.  I believe that nothing should persist from inside incognito mode to outside it.  There should be no history of browsed websites / URLs.  You should assume that the user is browsing something that could get them in trouble later.  For instance, a dissident could be returning from abroad to their home country after having read websites that are considered subversive in their home country.  This is a serious matter that should not to be taken lightly.  People's lives could be destroyed by this.

If you are going to save any such information, then there should be an explicit dialog each time something is about to be saved that will tell them exactly what will be stored and will allow them to cancel the action if they do not agree.

If I understand the current state of the plugin, there is no URL history or other browsing history stored unless you move one of this sliders or hit one of the buttons.  Is this correct?  In that case, would it be easy to trap all of those actions and pop up such a dialog?

**Comment by nekromoff** [comment]
Currently, PrivacyBadger remembers settings from normal browsing and uses
them when switched to Incognito as well. Based on the comments above this
should not be the case.

**Comment by tshanks** [comment]
I'm just insisting that it not save any history of browsed sites without explicit approval from the user each time.  Either the strategy of only saving them temporarily in memory during the current incognito session or the strategy of asking each time a change is requested would be fine with me.

I'm also asking if my understanding of the current behavior is correct.  My understanding is that PB doesn't save anything about URLs / domains from within incognito mode unless you click the badger and change something.  Is this correct?
